### ISSUE TITLE
Add additional metadata field for registering RI datasets / training / learning material used in a course or programme

### ISSUE THREAD

**Issue by vronk** [acdh-oeaw/dhcr-main#48]
Provide explicit definition of a "MOOC" and also other types of courses, maybe in form of a glossary

- [x] Discuss and formulate the definitions, also comparing MOOC and Online Courses
- [x] write and make available glossary
- [x] Introduce in db: as additional education type
      (or as attribute (would require change of datamodel))
- [x] corresponding changes in managing courses (adding the education type)
- [x] Adapt search options on "home page" `Choose Education`-filter

Edit on 2024-05-14 by Patrick:
The purpose of this issue has changed, please check here: https://github.com/acdh-oeaw/dhcr-main/issues/48#issuecomment-1724269045

**Comment by IvdL22** [comment]
Thank you @vronk for creating this task. I noticed that we already have a draft document with definitions started by @PixlTracer: https://docs.google.com/document/d/1FMquEoMn7EIOHH6I00K_518cDIVzBsvShUnKCut0FGE/edit?usp=sharing. We will finalise the glossary in that document and notify you when done.

**Comment by patrickakk** [comment]
@IvdL22  @PixlTracer :

Would it be possible to provide 2 examples of a MOOC? With 2 links to more detailed information? So that we can use that as reference when working on this topic?

**Comment by IvdL22** [comment]
@patrickakk cc @PixlTracer We now have a glossary here where you can check the terms and examples: https://docs.google.com/spreadsheets/d/1rXZQa-TsxTKXvyfZx4zeUxkC1_kqODO4/edit?usp=sharing&ouid=107443344257746111643&rtpof=true&sd=true

**Comment by patrickakk** [comment]
@IvdL22  @PixlTracer @vronk Summary from today's meeting:

It's no longer the intention to include MOOC's in the DHCR.

The **open educational resources** from Dariah Teach should be included. One example provided by Iulliana is:
https://teach.dariah.eu/course/view.php?id=68&section=2

Important information fields are: title, description and link.

The next step would be for Patrick to analyse the structure of the new items and see if and how we can integrate this in the registry.

Is that a correct summary?

Note: since this is additional work, I'll move this task to the mai23 milestone.

**Comment by IvdL22** [comment]
@patrickakk I agree with your summary and suggestion.

**Comment by IvdL22** [comment]
@patrickakk @PixlTracer I have tested this by adding one course to the registry: [https://dhcr.clarin-dariah.eu/courses/my-courses](https://www.google.com/url?q=https://dhcr.clarin-dariah.eu/courses/my-courses&sa=D&source=docs&ust=1683727793427619&usg=AOvVaw3yZejU4uwXl9nWyM3CVaRD)

You as admin can view it. It is not listed. 

Conclusion: the metadata needs to be more flexible in order to be able to include open-source educational courses.

Institution - I have created DARIAH-TEACH
City - not applicable
Country - not applicable
ECTS - 1
Start date - anytime
Duration - it could be 1 week or 1 month, depending on the learner
Discipline - I could not select DH so I selected Other
Technique - I did not know which one to select. Could we add Other as an option?
Objects - DH

**Comment by patrickakk** [comment]
Moved to July milestone since the specs are not finished. Maybe we need to move it again, depending on how long this takes.

**Comment by patrickakk** [comment]
@PixlTracer  @IvdL22  @vronk 
With this comment, I'll try to summarize the meeting from 2023-05-17 as well as propose some solutions:

## Characteristics

Open education resources could be used by teachers to include them into other courses. As well they could be used by students.

Question: We used various terms. Which one is correct? Is this: Open Education Resource (OER)?

Currently they can be found on multiple locations:
-DARIAH [ca 16 courses]: https://teach.dariah.eu/course/index.php 
-CLARIN [ca 10 courses]: https://www.clarin.eu/content/training-materials
-Upskills [ca 6 courses]: https://upskills.fil.bg.ac.rs/

When adding them to the DHCR, the metadata of all of them will be available at one location, as well as the complete overview can be accessed through the API.

An example for recent updated(public shown), PhD courses though the API can be found at: 
https://dhcr.clarin-dariah.eu/api/v2/courses/index?recent&course_type_id=4
The same could be available when a new education type "Open Education Resource" will be created.

It was proposed to enter the courses manually, since this is a small amount of courses and the metadata is not expected to change often. The information could be checked/updated yearly by the administrators (Anna & Iulliana). 

### The characteristics of the OER's conflict with the current data model. Three options were discussed:
1. Add one course which "summarizes" the OER's at one place. (Similar to ACDH tool gallery). In this case it might be difficult for a user to find the details.
2. Create a separate list/database/table/data model for the new characteristics. This is almost the same as creating a separate registry and requires a lot of development hours. As well the items won't be available though the current filter/search options.
3. Create a new Education Type and define a "set of rules" how we can deal with the differences in structure. This might be the preferred option, depending on how we can change/adapt the current validation rules.

### Which characteristics are different? OER's have:
-No physical present institution and department
-No city and country and no location on the map (lat, lon)
-No start date and is not recurring
-No fixed duration [unit and type] (that would depend on the student)
-No lecturer name and email = no problem, since those fields are not required
-No entry requirements? = no problem, since this field is not required
-Are there Tadirah objects which apply?

The goal is that the data model represents the real world. Currently the validation rules apply to all courses, which means there are no exceptions based on education type. For example every course has to be at an existing institution and that has to be on a physically existing location. Another example is that every course needs to have at least one start date. The presence of these data is also important when users are using the filter or sort options.

We could try to implement different validation rules, which behave different, based on the Education Type. In case we choose this solution, I'll have to take a look at this further, to see if the model supports this.


## What should the new validation rules do?

### Institution and location difference
- Create institutions for the OER materials. How many do we need?
	- Create validation rules which prevent not OER courses from using those institutions.
- Create city and country "Remote" for all OER courses
	- OER courses should be forced to use only the "Remote" location. Otherwise they will "pollute" the filter results when searching for a regular course.

### No start date available
- Option 1: Allow empty start date for OER's. 
    - Disadvantage: In the registry (list) they will look like courses which are *not* starting any more.
- Option 2: Set a start date for a fixed day in every month as well as check(enable) recurring. For example 1th of Jan, 1th of Feb, etc.
    Disadvantage: Based on start date (always very soon) they might get more attention than the currently existing courses
- Option 3: Something else?

### No fixed duration [unit and type]
We could use the same approach as for the location difference:
Create a new duration unit "Flexible" which is required for OER courses and can not be used by other courses.
Would that be a good solution?


## How can this new feature be communicated / be visible for users? How should OER courses be visible / findable for users?

1. The user can filter on the new education type "Open Education Resource". (Use filter button)

2. A section could be added in the menu, which contains an explanation of OER's as well as a button which automatically activates the filter and shows all the OER's. An example of clicking on this button, for the course type "PhD", would look like this: https://dhcr.clarin-dariah.eu/?course_type_id=4

3. A short url could be created, for easy dissemination, for example: https://dhcr.clarin-dariah.eu/open-education-resources 
Is this needed? Which short url is preferred?

4. Is anything else needed?

Is this summary correct? Did I miss anything? Are there suggestions/additions? And finally, what do you think of the proposed solution?

**Comment by vronk** [comment]
I have to admit, I am somewhat surprised about the turn this task took.
As the analysis also shows, the educational resources are ontologically quite a different animal than courses:
the former are some kind of digital object, the latter an activity.

And! For a catalogue of resources we have alone in DARIAH the DARIAH Campus: https://campus.dariah.eu/resources/page/1
besides many other catalogues and directories. Especially dariahTeach materials are also already included in DARIAH-Campus.

So I am decidedly against extending the functionality to support OER as first-class citizens next to courses.
In my recollection, the original idea was to allow for links/pointers to resources (learning materials) pertaining to existing course,
not a separate index of learning materials.

**Comment by PixlTracer** [comment]
Matej and me had a chat about this and we propose to have a call with Iulianna and Toma to discuss and find a suitable solution

**Comment by patrickakk** [comment]
Removed from milestone as discussed in meeting

**Comment by vronk** [comment]
As discussed today with Toma Tasovac (see [agenda & minutes](https://docs.google.com/document/d/11Tk00B-5pY1fMSSAHyzBspvWHi2Btsr2ovt2SoHODUs/edit)):

We want to keep a clear distinction between courses
as **Activities** (event with start and end date and a location) and
as **Digital objects** (like training materials available online).

And restrict DHCR to register and present the "Activities" and use DARIAH Campus for the "Digital objects".

That means that "online courses" available e.g. in dariahTeach or from the upskills project should not be registered in DHCR, but rather in D-C as external resources.(Indeed courses from dariahTeach already are featured in D-C.)

However, it should be possible to record, if there is training material available accompanying the corresponding activity. This should be in form of a field to be populated ideally with a URL pointing to the online material (ideally hosted in a stable manner). Indeed such material can be made available through DARIAH Campus, in which case the entry in DHCR would point to the corresponding entry in D-C.

**Comment by patrickakk** [comment]
@vronk Thanks for the explanation.

@vronk @PixlTracer @IvdL22 Can we close this issue? If yes, please change the label to "Done" or click on Close ;)

**Comment by PixlTracer** [comment]
Since we agreed to introduce a metadata field, where the course owner/lecturer can leave a link that points to the training/learning material available (see @vronk's comment above), I'd like to leave this issue open, until the metadata field has been implemented;

--> to do @patrickakk:
introduce a (non-mandatory) metadata field in which a link can be set that points to learning material hosted/stored elsewhere

proposal:
when adding a course via https://dhcr.clarin-dariah.eu/courses/add, I propose to insert the following metadata field:

![dhcr new metadata field for course material](https://github.com/acdh-oeaw/dhcr-main/assets/117655705/a8909ee7-c9da-4fe0-b8f3-ff447975a9b1)

@IvdL22 please review the wording and indicate changes/give green light.

thank you!

**Comment by patrickakk** [comment]
@PixlTracer Thank you for pointing that out.

Can we specify what is needed:
- One new text field (as shown on the screenshot above)?
- Are we sure there will always be one or no link? This means not more than 1? Does that apply in all cases?
- Should there be a simple check if it contains an url? (http)
- Should there be a url validation check?
- Should that only be done on entry/update or also on regular basis?

(For now I moved it to the November milestone. After the specifications are clear, we could talk about the milestone for the implementation?)

**Comment by PixlTracer** [comment]
> @PixlTracer Thank you for pointing that out.
> 
> Can we specify what is needed:
> 
> * One new text field (as shown on the screenshot above)? **_yes_**
> * Are we sure there will always be one or no link? This means not more than 1? Does that apply in all cases? **_we will not know. we could offer a 2nd link field (which might not be used much)..._**
> * Should there be a simple check if it contains an url? (http) **_yes, good point! check for https://_**
> * Should there be a url validation check? **_absolutely!_**
> * Should that only be done on entry/update or also on regular basis? **_what do you mean with 'regular basis'? url validation check every now and then? -- would be useful, but how to handle invalid links then? remove? contact course owner? (question also for @IvdL22)_**
> 
> (For now I moved it to the November milestone. After the specifications are clear, we could talk about the milestone for the implementation?)

**Comment by patrickakk** [comment]
Preview of user interface with dummy data.

# Public available info at Course Detail page

![image](https://github.com/acdh-oeaw/dhcr-main/assets/97846194/9fa98540-7b69-4055-bceb-74a78fd1de77)

# New field in login area at Course add or edit

![image](https://github.com/acdh-oeaw/dhcr-main/assets/97846194/d56155c8-ba6f-4da1-8570-0e27eb16fcc0)

**Comment by patrickakk** [comment]
@IvdL22 
cc @PixlTracer 
Is this the correct summary of what was said in the meeting on the 10th of June?

CLARIN needs to possibility that links (=plural) to datasets used in a course, are added to the course metadata.
DARIAH needs to possibility that links (also plural?) to training material that is (re)used in a course are added to the course metadata.

On the other hand, based on the requirements here:
https://github.com/acdh-oeaw/dhcr-main/issues/48#issuecomment-1755445882

> Are we sure there will always be one or no link? This means not more than 1? Does that apply in all cases? we will not know. we could offer a 2nd link field (which might not be used much)...

and

> Should there be a url validation check? absolutely!


The current (almost finished) implementation only contained one text field, with place for one link and a link checker that required a valid http status code. So it's not possible to enter more than one link.

The information above almost certainly needs more than one item to be added, which should be implemented in the data-model in a completely different way. (One to many relationship).

Based on the new information above the current implementation does not provide what is needed.

During the meeting it was decided to discuss what is needed during the WG meeting on the 19th June.

I'll revert and not commit the work already done, put the issue on hold and move it to the July milestone. Maybe we can all  agree on what is needed soon?

**Comment by IvdL22** [comment]
Hi @patrickakk cc @PixlTracer Thank you.  The way I see this implementation is the following:

Description: If you use a CLARIN or DARIAH resource in your course, be it training material, dataset, tool or service, please add the URL.

Optional field 1: CLARIN resource 

Option field 2: DARIAH resource 

What do you think? Would this be possible to implement? I could add the text like this to the slide and then discuss with the WG.

**Comment by patrickakk** [comment]
@IvdL22 @PixlTracer

I would suggest to go a (few) steps back in the process: Why do we develop this feature?

Is this because CLARIN and DARIAH want to know which resources are used and in which courses or how often they are used? In that case you want to measure something, which makes it an important feature?
And in that case, why do you want to ask during a WG meeting what they need?

What do we do when there is more than one link to add?

In the case of a CLARIN dataset, do you expect they always only used one?
What do you do when they used more than one and can't enter the information? Then your report is unreliable?

Do we need to add attributes to the link? (Source: DARIAH/CLARIN, Type: Dataset, tool, training material, etc.)?

How do you want to see/summarize the data which is entered?

With the requirements/wishes currently available, I would suggest the following steps:

a) Iulliana and Anna provide at least three(3) real world examples of courses and the links that can be added

b) We create a dummy preview of a report for both organizations and check if that contains the information that's needed, in the format/with the attributes needed.

c) Based on point B, we decide which information, attributes and data structure is needed.

d) Dummy user interface preview. Now, this needs to be accepted before proceeding to the next step.

e) Decide if the new fields should be provided by the API as well.

f) Technical implementation

This is probably not a small feature. Since the larger amount of working hours needed for this, I would suggest to have at least 1 dedicated meeting about this, where at the end, everybody commits to a set of requirements to avoid doing the same work over and over again.
And at least involve Matej @vronk as well.

And please think about this: It's easy to make changes to the wishes now, more complicated when the feature is finished and way more complicated when some data has been entered. So what do you need from this feature in 1 year from now?
A small misunderstanding now, can cause a lot of additional working hours later.

**Comment by vronk** [comment]
The idea is that there can be more than 1 URL reference for datasets/training material (or other resources), no distinction between CLARIN and DARIAH

a separate 1:N table: **external_resources**

```
[<“label”,”URL”,”type”, “affiliation”>,..] 
```

“label” is an open text describing the resource (optional)
“type”= Dataset, Training Material, Service, Software, …(optional)
“affiliation” = CLARIN, DARIAH, … (optional)
{label}/{URL}

```
{affiliation}{type}: {label}
{URL}
```

> CLARIN Dataset: CLARIAH.NL corpus of child speech
> [https://clariah.nl/…](https://clariah.nl/%E2%80%A6) 

Link checking upon submission (optionally if little effort)
accept HTTP Status >= 200 < 400

The new fields should be available via the API too.

**Comment by vronk** [comment]
@IvdL22, @PixlTracer could you please provide at least three real world examples of such external resources according to the proposed data structure

**Comment by IvdL22** [comment]
@vronk @patrickakk @PixlTracer Matej, thanks for the technical solution.

Example
This is a course in the registry:
[Puheen analyysin perusteet (Introduction to Speech Analysis)](https://dhcr.clarin-dariah.eu/?sort=Courses.name:asc,Countries.name:asc,Institutions.name:asc,CourseTypes.name:asc#550)

“label” Introduction to Speech Analysis
“type”=Training Material
“affiliation” = CLARIN 
{label}/{URL] https://www.clarin.eu/content/introduction-speech-analysis

“label” Route to a Wing Corpus
“type”= Dataset
“affiliation” = FIN-CLARIN
{label}/{URL] http://urn.fi/urn:nbn:fi:lb-2020112929

The same course also uses the CLARIN VLO to show students how to search for other corpora, so I could also add the service.

Is this example enough?

Edited by patrickakk on 2024-07-03. Reason: Fixed links

**Comment by patrickakk** [comment]
@IvdL22 @PixlTracer 
cc @vronk 
Thank you for providing one example. Would it be possible to provide the other two examples as well?

Could you include at least one example which uses DARIAH resources? And in general: include as much exceptions and complicated situations as possible?

@IvdL22 
Could you also specify how the VLO should be added to the example you provided?

**Comment by IvdL22** [comment]
@patrickakk 
cc @vronk @PixlTracer 
Regarding the VLO, if a teacher is using a corpus found via the VLO, the labels will be the same:

“label” Nijmegen corpora of casual speech
“type”= Dataset 
“affiliation” = CLARIAH-NL
{label}/{URL] https://hdl.handle.net/1839/2581a242-fde3-4349-ab06-4920a964803d

Regarding the DARIAH Campus: if a teacher is using a learning or training resource in the DH programme entered in the registry, the teacher should be able to add the link to the resource:

“label”Formal Ontologies: A Complete Novice's Guide
“type”=Training Material
“affiliation” = DARIAH CAMPUS
{label}/{URL] https://campus.dariah.eu/resource/posts/formal-ontologies-a-complete-novices-guide

Please let me know if some things are not clear. Thank you for your help.

**Comment by patrickakk** [comment]
@IvdL22 
cc @PixlTracer @vronk 

Thank you for the examples. Do you agree, when considering the specifications here:
https://github.com/acdh-oeaw/dhcr-main/issues/48#issuecomment-2167415842
that the affiliation in both last examples should be "CLARIN and "DARIAH" ?

Since according to :

> “affiliation” = CLARIN, DARIAH, … (optional)

There are only two options for this value. The idea behind this was that a resource belongs either to CLARIN or to DARIAH (or there is an exception).

Is there a special reason for specifying this more detailed?

This is an important difference, please let me know if you think otherwise or have any questions.

**Comment by IvdL22** [comment]
@patrickakk Both CLARIN and DARIAH have national consortia and repositories, so the affiliation can also be CLARIN-FIN, etc. As @vronk suggested, there could be more values, not only two. The course contributors should be free to fill the institution hosting the resource they use in teaching.

**Comment by patrickakk** [comment]
@IvdL22 
cc @PixlTracer 

Thank you for the explanation and valuable insights. I was aware of the national consortia, but until now it's wasn't clear to me that you wanted to specify on that detail level.
Maybe situations like this are also a good example for everbody, why it is useful to ask for examples and ask such questions in this (early) stage of the development process?

1. Type of values
You also mentioned "course contributors should be free to fill..." To avoid any misunderstandings about that, I will start with describing that kind of input that was meant and as second step we could discuss the values.

> [<“label”,”URL”,”type”, “affiliation”>,..] 
> “label” is an open text describing the resource (optional)
> “type”= Dataset, Training Material, Service, Software, …(optional)
> “affiliation” = CLARIN, DARIAH, … (optional)

- Label will in the UI look like a text field, where the course contributor can enter a free chosen value. The field is optional.
- Url will in the UI look like a text field, where the course contributor can enter a free chosen value. The field is mandatory.
- Type will in the UI look like a drop-down menu with a list of pre-defined values. The field is optional.
- Affiliation will also be a drop down menu with a pre-defines list of values. How else would you prevent pollution or similar items/duplicates? The field is optional.

Does that clarify a bit?

2. Kind of affiliation
As far as I understood, the idea was to let them choose from either CLARIN, DARIAH, or none.
Now you mentioned it should be possible to specify the  national consortia.
If we change this how should it be possible to answer for example the following questions, using the dataset:
-Which courses use a DARIAH resource?
-Which courses are in the german language and use a CLARIN resource?

A possible solution could be to define an affiliation parent type. Values for this could be DARIAH or CLARIN?
And define the national consortia as affiliation child types.
Are we sure that every national consortia belongs to only one parent type?
And of course this makes the implementation more complex, which means more working hours are needed.

On the other hand, the example provided by vronk shows something different:

> CLARIN Dataset: CLARIAH.NL corpus of child speech
[https://clariah.nl/…](https://clariah.nl/%E2%80%A6) 

affiliation = clarin
type = dataset
label = CLARIAH.NL corpus of child speech

Are we still all on the same page?
Is this understandable or should we discuss this in a (short) meeting?

**Comment by IvdL22** [comment]
@patrickakk , if the example by @vronk is clear and you know how to implement it, why do you ask me for more examples? :)
@vronk Matej, could you please help us move forward with this task? I provided examples but I am not able to clarify the specifications. Thank you :)

**Comment by patrickakk** [comment]
@IvdL22 To make sure that:
a) What we defined, matches what you need
b) To check if there are any misunderstandings
c) Making sure that we all have the same expectations
d) Avoid wasting valuable development hours, by checking those things before starting the work

**Comment by vronk** [comment]
have a flat list (modifiable by administrators) containing both CLARIN and national consortia.
allow multi-value in the field

example:
affiliation: ['CLARIN', 'FIN-CLARIN']

**Comment by vronk** [comment]
as for discussion strategic meeting: this would be high priority

**Comment by patrickakk** [comment]
as discussed in the meeting today. updated specs:

No list of detailed affiliations.
two values: CLARIN, DARIAH, empty or both !!
=> still separate table with N:N relations
editing functionality optional

**Comment by vronk** [comment]
Suggestion to:

* add a button from external resources page to [Back to editing course] 
* add a button to main course page to go optionally to external resources: [Update course & Add resources]

**Comment by vronk** [comment]
for the frontend rendering:
* Heading "External Resources" - make it a heading as Description (H3?)
* list the resources as bullet list
* remove the gap between the title and the url

**Comment by patrickakk** [comment]
The feature is available to be tested, in the test instance, here:
https://test-dhcr.clarin-dariah.eu

Please note: **The test instance uses a separate database. 
This allows to experiment and test. Persistent changes should be made in production.**

It contains the following parts and changes:
### Course list

- Changed Update/Edit to Update. 
- Added Ext. Resources as direct link to manage the External Resources per course.

![Image](https://github.com/user-attachments/assets/460bb6d6-2c9e-4623-b6fa-d48c31373d3f)

### Updated flow on course add and edit page

![Image](https://github.com/user-attachments/assets/c85f1942-3572-4898-9bc8-532058854aad)

### New section: "Show External Resources"
(After clicking on second button in screenshot above)

![Image](https://github.com/user-attachments/assets/1cb81e25-c87d-4bfa-b27f-3a820950bc59)

### Add external resource

![Image](https://github.com/user-attachments/assets/c62a8078-896a-41dc-81f0-8817527cfe7f)

### URL validation

- Valid URL format
- Start with https:// or http://
- This is **not** a link checker

![Image](https://github.com/user-attachments/assets/70611a59-9b9a-4abd-97be-05ac6d48517c)

### Added affiliation "CLARIN & DARIAH". As mentioned in the requirements as "both".
See bottom part of screenshot above.

### Edit external resource

![Image](https://github.com/user-attachments/assets/a157a01f-36ed-4c8b-b09a-5a5bb6b25dc1)

### Overview

![Image](https://github.com/user-attachments/assets/a0f53668-1b09-450f-b2a8-41bfe2860beb)


### Public view (course detail page)
https://test-dhcr.clarin-dariah.eu/courses/view/550

![Image](https://github.com/user-attachments/assets/b7aea53d-e1ca-4962-a61c-996b96274fc3)

**Comment by patrickakk** [comment]
Example:
https://test-dhcr.clarin-dariah.eu/courses/view/814

![Image](https://github.com/user-attachments/assets/269e83aa-d974-4a7a-9177-c487dcdf38d5)

**Comment by IvdL22** [comment]
@patrickakk @vronk I am thrilled to see the implementation of this feature! Congratulations!!!! I just tested it and it works very well. It is very easy to add external resources! 

@patrickakk 
- One suggestion to change Url* into URL*.
- Would it be possible to add the following description under **Show External Resources**? https://test-dhcr.clarin-dariah.eu/external-resources/show-ext-resources/763

"As of May 2025, the course contributors can add CLARIN and DARIAH external resources they use in their course/programme/summer school. An external resource can be "Dataset, Training Materials, Service, or Software". Course contributors can add as many resources as they need".

@patrickakk @vronk Is there a limitation for the nr of resources the users can add? If yes, I can change the text.

**Comment by patrickakk** [comment]
Another example:
![Image](https://github.com/user-attachments/assets/e1b92e64-aa3c-4b3f-a544-6da8a2d39c06)

![Image](https://github.com/user-attachments/assets/d75169a4-4c9c-4cb9-9f8a-29e6f8d7f315)

![Image](https://github.com/user-attachments/assets/e24057e0-cd38-4699-a232-b934319182d6)

![Image](https://github.com/user-attachments/assets/6a970e3a-76b2-4407-b842-673f8d855cc2)

![Image](https://github.com/user-attachments/assets/c7f2c061-4c26-4826-afb8-d591e6cfd97c)

**Comment by patrickakk** [comment]
@IvdL22 
Thank you. That is always nice to read.
The changes you've requested are implemented.
Unfortunately deployment currently doesn't work, so there is nothing online.
The screenshots below are taken from my laptop to give you a preview.

Screenshots for Iulliana:

# Updated text URL*

![Image](https://github.com/user-attachments/assets/8291337a-3bea-4d13-97c8-71d277828a92)

# Added explanation
@IvdL22 I've changed Training MaterialS to Training Material, to stay consistent with the wording used at the add/edit page and as well write everything in singular. Do you agree with that?

![Image](https://github.com/user-attachments/assets/24d0548c-e87b-4e98-9ede-98e06e1aed1f)

To answer your last question:
There is no limitation on the amount of external resources that can be added.
Maybe only the amount that is actually used is a course ;)

Did this answer/solve all your questions?

**Comment by patrickakk** [comment]
The changes above are implemented in 2025-05/2

**Comment by patrickakk** [comment]
Actions postponed for later implementation:

- [x] Create overview (index) page of all external resources, in login area for the administrators 
- [x] Add counts to statistics dashboard
- [x] Fix updated field is not filled with a value
- [ ] Fix existing unit test and enable the temporary disabled existing unit tests
- [ ] Write units tests for new added funtions
- [x] Extend API to include the full list of all external resources

**Comment by IvdL22** [comment]
@patrickakk Thank you for the screenshots, I agree with the wording. Looking forward to the deployment, many thanks

**Comment by patrickakk** [comment]
# Overview of all external resources 
(for admins only)

### New menu entry

![Image](https://github.com/user-attachments/assets/e16e6acd-ab8e-4a0c-8d5f-a35b4f7fcffa)

### New dashboard item

![Image](https://github.com/user-attachments/assets/89b146ea-dc18-4582-8f93-9b4fa051958a)

### Overview page

![Image](https://github.com/user-attachments/assets/b23f56dc-6c52-42a7-bfd6-bdc0e23e7f2b)

**Comment by patrickakk** [comment]
# Added counts to summary statistics dashboard

![Image](https://github.com/user-attachments/assets/7d2964d3-040d-4781-b39a-fdd3515eb073)

![Image](https://github.com/user-attachments/assets/acdf38c0-83e3-4107-b8a1-3dd2a8cbe45a)


Feel free to suggest other calculations.

Note: These statistics are about dummy data. Please use the production instance (in the future) for accurate and live statistics.

**Comment by patrickakk** [comment]
@IvdL22 @PixlTracer 

Can you review the two new features, as specified above in this issue?

- Overview of all external resources
- Added counts to summary statistics dashboard


Both are implemented in 2025-06

**Comment by vronk** [comment]
As discussed, we don't think we need to have a full list of all resources.

**Comment by PixlTracer** [comment]
Contrary to @vronk's comment, my humble opinion: I do think it is a neat feature, to see at one glimpse, how many courses do have an external resource assigned. (we can also monitor this way whether the community makes use of this new feature.)
however the metrics "courses with at least one" and "avg per course when at least one" are to be questioned, whom do they serve? (question may remain unanswered)

to sum up: I am happy with "total" and "published", but do not see the sense of  "courses with at least one" and "avg per course when at least one", the latter two metrics can be removed - @IvdL22 ok for you?

thank you @patrickakk for this new feature, the course statistics page is quite fancy now!

**Comment by patrickakk** [comment]
@PixlTracer 
cc @IvdL22  @vronk 

There seems to be a misunderstanding:

The comment

> As discussed, we don't think we need to have a full list of all resources.

Is about the task "Extend API to include the full list of all external resources" which is listed here:

https://github.com/acdh-oeaw/dhcr-main/issues/48#issuecomment-2907807719

In the meeting at 2nd July it was decided: There is no specific API endpoint needed for the external resources. They are listed as child-elements in the course-view API endpoint.

Then Matej wrote the comment and I've marked the task as completed. This is independent from the two new small features which were presented above.

To respond to your question:

The current layout of the statistic page doesn't provide much space for longer text explanations. Therefore it has been summarized  in a few words.

You mentioned

> how many courses do have an external resource assigned

That is exactly what the number at "courses with at least one" shows.

The first two metrics are a sum of the amount of external resources.

The second two metrics show the relation between external resources and courses.

Full explanation:

1. Total amount of external resources entered
2. Same as 1, only published ones
3. Total number of courses with at least one external resource assigned. (doesn't take into account if a course is outdated, but that's not relevant now)
4. Average amount of external resources per course where at least one external resource is assigned.

For me both point 3 and 4 are the most valuable metrics, since they show how this feature is accepted and used by the community.

My suggestion would be to keep the existing metrics. Since they are already there, don't hurt anyone ;) and this is accessible by admins only.

Nice to read that you are also happy with the new features.

**Comment by patrickakk** [comment]
@PixlTracer @IvdL22 

Can you let know if you agree?

**Comment by patrickakk** [comment]
Last two remaining tasks:

- [x]  Fix existing unit test and enable the temporary disabled existing unit tests
- [ ]  Write units tests for new added funtions

**Comment by IvdL22** [comment]
@patrickakk I agree with your suggestions about point 3 and 4 above. Thank you!

**Comment by patrickakk** [comment]
@IvdL22 @PixlTracer 

That's great. I took the initiative of changing the label to done.

The existing unit tests are also fixed and enabled again.
Specific unit tests for this feature have not been created. In the last meeting it was also mentioned that is has not priority.

So we can finally say: Case closed!  ;)

Implemented in 2025-08
